qiskit/circuit/classicalfunction/boolean_expression.py

```diff
@@ -0,0 +1,173 @@
+                              '"pip install tweedledum".')
+        from tweedledum import BoolFunction
+        self._tweedledum_bool_expression = BoolFunction.from_expression(expression)
+
+        short_expr_for_name = (expression[:10] + '...') if len(expression) > 13 else expression
+        super().__init__(name or short_expr_for_name,
+                         num_qubits=self._tweedledum_bool_expression.num_inputs() +
+                                    self._tweedledum_bool_expression.num_outputs(),
+                         params=[])
+
+    def simulate(self, bitstring: str) -> bool:
+        """Evaluate the oracle on a bitstring.
+
+        This evaluation is done classically.
+
+        Args:
+            bitstring: The bitstring for which to evaluate.
+
+        Returns:
+            bool: result of the evaluation.
+        """
+        from tweedledum import BitVec
+        input = []
+        for bit in bitstring:
+            input.append(BitVec(1, bit))
+        return bool(self._tweedledum_bool_expression.simulate(*input))
+
+    def synth(self, registerless=True):
+        """Synthesis the logic network into a :class:`~qiskit.circuit.QuantumCircuit`.
+
+        Args:
+            registerless (bool): Default ``True``. If ``False`` uses the parameter names to create
+            registers with those names. Otherwise, creates a circuit with a flat quantum register.
+
+        Returns:
+            QuantumCircuit: A circuit implementing the logic network.
+        """
+        from tweedledum.passes import pkrm_synth
+
+        if registerless:
+            qregs = None
+        else:
+            qregs = self.qregs
+
+        logic_network = self._tweedledum_bool_expression._logic_network
+
+        return tweedledum2qiskit(pkrm_synth(logic_network), name=self.name, qregs=qregs)
+
+    def _define(self):
+        """The definition of the boolean expression is its synthesis"""
+        self.definition = self.synth()
+
+    @classmethod
+    def from_dimacs_file(cls, filename: str):
+        """Create a BooleanExpression from the string in the DIMACS format.
+        Args:
+            filename (str): A file in DIMACS format.
+
+        Returns:
+            A quantum circuit (BooleanExpression) for the input string
+        """
+        if not HAS_TWEEDLEDUM:
+            raise ImportError("To use the BooleanExpression compiler, tweedledum "
+                              "must be installed. To install tweedledum run "
+                              '"pip install tweedledum".')
+        from tweedledum import BoolFunction
+
+        expr_obj = cls.__new__(cls)
+        expr_obj._tweedledum_bool_expression = BoolFunction.from_dimacs_file(filename)
+
+        super(BooleanExpression, expr_obj).__init__(
+            name=basename(filename),num_qubits=expr_obj._tweedledum_bool_expression.num_inputs() +
+                                               expr_obj._tweedledum_bool_expression.num_outputs(),
+            params=[])
```

Comment:
Alternatively, would it be more straightforward to do the following?
```python
expr = BoolFunction.from_dimacs_file(filename)
return cls(expr)
```

Reply:
this requires `from_dimacs_file` to create an expression, which needs to be parsed by constructor. I agree this `from_<somewhere>` is not nice.